multiregionccl: add global tables roundtrip test

arulajmani opened the pull request:
This patch adds TestEnsureLocalReadsOnGlobalTables. which ensures that
global tables serve present time reads locally.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64387)
<!-- Reviewable:end -->

nvb reviewed (commented):
Cool test!

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-:-MZTKgopEqXTDXfFz0eD:byo8gf1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @otan)



<!-- Sent from Reviewable.io -->

otan reviewed (approved):
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @arulajmani)

___
*[pkg/kv/test_utils.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-MZUYa9aDDorSfO7Hbxx:-MZUYa9aDDorSfO7Hbxy:b2jkay7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b573122836fe36e774ec332ef492df1a502e0a6e/pkg/kv/test_utils.go#L45)):*
> ```Go
> 		if sp.Operation == "dist sender send" {
> 			if tracing.LogsContainMsg(sp, kvbase.ServingLocalReadMsg) {
> 				foundLocalReads = true
> ```

wondering if it is worth returning an error with the non-local read


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
I agree, very cool!  In a subsequent PR, is there some way to extend this test (or create a new one) to cover concurrent writes on GLOBAL tables to ensure that they don't force a remote read (or wait time)?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-:-MZXpkkz5yLVvCWYkeFR:bk66ocb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @arulajmani)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> In a subsequent PR, is there some way to extend this test (or create a new one) to cover concurrent writes on GLOBAL tables to ensure that they don't force a remote read (or wait time)?

I changed this test to add some background write traffic. I also changed what we're looking for in the spans slightly after discussing with @aayushshah15 offline -- now we check that the read was routed to the local node _and_ non-leaseholders served a follower read. Aayush also made me realize the `OnlyLocalReads` method wasn't working as intended for more complex spans, so I've made it local to the test. RFAL 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-:-MZ_aPKAC3WY8pguxTsw:b-e9efyo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @arulajmani, @nvanbenschoten, and @otan)



<!-- Sent from Reviewable.io -->

otan commented on pkg/ccl/multiregionccl/roundtrips_test.go:
```diff
@@ -0,0 +1,135 @@
+		followerRead := ensureOnlyLocalReads(t, rec)
+
+		if followerRead {
+			numFollowerReads++
```
can we raise a stronger assertion as to which nodes are global reads?

otan commented on pkg/ccl/multiregionccl/roundtrips_test.go:
```diff
@@ -0,0 +1,135 @@
+				_, err = sqlDB.Exec(`INSERT INTO t.test VALUES($1)`, i)
+				i++
+				if err != nil {
+					errCh <- err
```
if there is an error here, this blocks until the whole numServer routine completes. is that expected? Should we be using select somewhere below to consume the errCh?

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @arulajmani, @nvanbenschoten, and @otan)

___
*[pkg/ccl/multiregionccl/roundtrips_test.go, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-MZTI5RNopbJa3vUSvnK-r3-92:-M_LilV61EoRwn8tndFk:b-t6cmfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ceb9f622e38a883a2c1756223da8d8f72f450a/pkg/ccl/multiregionccl/roundtrips_test.go#L92)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

if there is an error here, this blocks until the whole numServer routine completes. is that expected? Should we be using select somewhere below to consume the errCh?
</blockquote></details>

I think it's just easier to return if there's an error here. 
___
*[pkg/ccl/multiregionccl/roundtrips_test.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-MZTI5RNopbJa3vUSvnK-r3-124:-M_Ljjec21bE_wJGvjnv:b-x5rr8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ceb9f622e38a883a2c1756223da8d8f72f450a/pkg/ccl/multiregionccl/roundtrips_test.go#L124)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can we raise a stronger assertion as to which nodes are global reads?
</blockquote></details>

How does this look? Expect non-leaseholders to serve a follower read?


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @otan)

___
*[pkg/ccl/multiregionccl/roundtrips_test.go, line 107 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-M_LnpqREYuL-IbVrs6n:-M_LnpqREYuL-IbVrs6o:bvklmkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e15deb53e70c08d6aae989d2203555293cd94d8/pkg/ccl/multiregionccl/roundtrips_test.go#L107)):*
> ```Go
> 	// Sleep to ensure all replicas have had the opportunity to apply a snapshot,
> 	// and are therefore primed to serve the follower reads we expect below.
> 	time.Sleep(kvserver.RaftLogQueuePendingSnapshotGracePeriod)
> ```

This isn't really doing what you expect. It is totally possible for this grace period to expire and for the range to not have upreplicated yet. The thing you want to check is whether all the 3 servers have a replica of the range. 

You could consider either using `ReplicationMode: ReplicationManual` and then manually add replicas to the other 2 stores, or you could poll all of the stores for a replica of the range like we might've briefly discussed. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @otan)

___
*[pkg/ccl/multiregionccl/roundtrips_test.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-MZTI5RNopbJa3vUSvnK-r3-124:-M_MuZLw-0gu_91OUqDl:b-sm3z9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ceb9f622e38a883a2c1756223da8d8f72f450a/pkg/ccl/multiregionccl/roundtrips_test.go#L124)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

How does this look? Expect non-leaseholders to serve a follower read?
</blockquote></details>

yes, but completely optional.
___
*[pkg/ccl/multiregionccl/roundtrips_test.go, line 133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-M_MuWzX0WVbh3TQRmIB:-M_MuWzX0WVbh3TQRmIC:byoevob) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ceb9f622e38a883a2c1756223da8d8f72f450a/pkg/ccl/multiregionccl/roundtrips_test.go#L133)):*
> ```Go
> 
> 	close(stopWritesChan)
> 	writeErr := <-errCh
> ```

this only consumes one item from the channel - there are more goroutines waiting.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/ccl/multiregionccl/roundtrips_test.go, line 133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-M_MuWzX0WVbh3TQRmIB:-M_N9_OoENbuCV3s7Fv7:b-d1nztk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8ceb9f622e38a883a2c1756223da8d8f72f450a/pkg/ccl/multiregionccl/roundtrips_test.go#L133)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this only consumes one item from the channel - there are more goroutines waiting.
</blockquote></details>

oh nvm i misread your PR


<!-- Sent from Reviewable.io -->

arulajmani committed:
sql: add options to TestingCreateMultiRegionCluster

Release note: None

arulajmani committed:
multiregionccl: add global tables roundtrip test

This patch adds TestEnsureLocalReadsOnGlobalTables. which ensures that
global tables serve present time reads locally.

Release note: None

otan reviewed (approved):
LGTM if it LGT-aayush

> Reviewed 6 of 12 files at r5, 3 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-:-M_Nm-0xBMUSoZktmMn8:bvp9vum)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @dt, @nvanbenschoten, and @otan)



<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64387#-:-M_OBfFZERB7qikqcDDF:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @dt, @nvanbenschoten, and @otan)



<!-- Sent from Reviewable.io -->

arulajmani commented:
merge skewed, trying agin. 

bors r=aayushshah15,otan
